flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/source/split/SplitsGenerator.java

```diff
@@ -0,0 +1,237 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.cassandra.source.split;
+
+import org.apache.flink.annotation.VisibleForTesting;
+
+import com.datastax.driver.core.ResultSet;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This class generates {@link CassandraSplit}s based on Cassandra cluster partitioner and cluster
+ * statistics. It estimates the total size of the table using Cassandra system table
+ * system.size_estimates. But there is no way to estimate the size of the data with the optional SQL
+ * filters without reading the data. So the splits can be smaller than {@param maxSplitMemorySize}
+ * when the query is executed.
+ */
+public final class SplitsGenerator {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SplitsGenerator.class);
+    private static final int ACCEPTABLE_NB_SPLIT_PARALLELISM_RATIO = 10;
+
+    private final CassandraPartitioner partitioner;
+    private final Session session;
+    private final String keyspace;
+    private final String table;
+    private final int parallelism;
+    @Nullable private final Long maxSplitMemorySize;
+
+    public SplitsGenerator(
+            CassandraPartitioner partitioner,
+            Session session,
+            String keyspace,
+            String table,
+            int parallelism,
+            Long maxSplitMemorySize) {
+        this.partitioner = partitioner;
+        this.session = session;
+        this.keyspace = keyspace;
+        this.table = table;
+        this.parallelism = parallelism;
+        this.maxSplitMemorySize = maxSplitMemorySize;
+    }
+
+    /**
+     * Split Cassandra tokens ring into {@link CassandraSplit}s containing each a range of the
+     * Cassandra ring of {@param maxSplitMemorySize}. If {@param maxSplitMemorySize} is not defined,
+     * or is too high or too low compared to the task parallelism, then it generates as many {@link
+     * CassandraSplit}s as the task parallelism.
+     *
+     * @return list containing {@code numSplits} CassandraSplits.
+     */
+    public List<CassandraSplit> generateSplits() {
+        long numSplits;
+        if (maxSplitMemorySize != null) {
+            final long estimateTableSize = estimateTableSize();
+            LOG.debug("Estimated table size for table {} is {} bytes", table, estimateTableSize);
+            numSplits = estimateTableSize / maxSplitMemorySize;
+            if (numSplits == 0 // estimateTableSize can be null in some cases (see javadoc)
+                    || numSplits < parallelism / ACCEPTABLE_NB_SPLIT_PARALLELISM_RATIO // too low
+                    || numSplits
+                            > (long) parallelism
+                                    * ACCEPTABLE_NB_SPLIT_PARALLELISM_RATIO) { // too high
```

Comment:
Well, off course this ratio is totally arbitrary. I just sketched a quick code to serve as a base discussion on safety measures for the number of splits. I agree with your arguments on limit cases (freedom for the user and dangerousness of sudden threshold crossing). 

I'll implement the lazy splits.

Reply:
> what if the user defined value is way above the table size? We fallback to one split ?

Since it's only a _maximum_ and not a _target_ size, so long as we are below the maximum we can generate as many splits as we'd like (== parallelism I suppose).

That's how I think about it at least.